### PR TITLE
cli: add \statement-diag SQL shell command

### LINKED ISSUE #70931: sql: EXPLAIN ANALYZE (DEBUG) not supported for non-system tenants
Non-system tenants may still wish to collect a statement bundle for debugging purposes using `EXPLAIN ANALYZE (DEBUG)`.
This is not yet possible.

### PR DESCRIPTION
**Pull request by RaduBerinde**
This PR adds a `\statement-diag` command to the SQL shell used to list and download bundles, and reenables `EXPLAIN ANALYZE (DEBUG)` for tenants.

#### cli: extract statement diagnostics code

This commit extracts the part of the statment-diag code that interacts
with the system tables and moves it to sqlclientconn.

Release note: None

#### cli: improve statement-diag download

Minor improvement to `statement-diag download`: make the filename
optional, for consistency with the SQL shell equivalent.

Release note: None

#### cli: simplify invalid syntax paths in SQL shell

This change simplifies most invalid syntax paths while improving the
error message in a few cases.

Release note: None

#### cli: add \statement-diag SQL shell command

This commit adds a `\statement-diag` command to the SQL shell which
can be used to list and download statement bundles directly from the
shell.

This command should be useful for multitenant (where the possibilities
of downloading the bundle are more limited), as well as running demo
with a `make buildshort` binary.

The output of `EXPLAIN ANALYZE (DEBUG)` includes the exact syntax.

Informs #70931.

Release note (cli change): the SQL shell now supports a
`\statement-diag` command for listing and downloading statement
diagnostics bundles.

#### sql: enable EXPLAIN ANALYZE (DEBUG) for tenants

EXPLAIN ANALYZE (DEBUG) was disabled for tenants because the returned URL
was unusable.

We can now download bundles through the SQL shell, so we can reenable
EXPLAIN ANALYZE (DEBUG). For non-system tenants, we hide the URL and
only show the SQL shell and cli options.

As part of this change, we also add tenant support to the interactive
cli test infrastructure and we implement the `--background` and
`--pid-file` flags for `mt start-sql`.

Fixes #70931.

Release note (sql change): EXPLAIN ANALYZE (DEBUG) can now be used by
tenants.

### CODE REVIEW

**Review by rytaft — APPROVED** [review]
:lgtm: (I'm not familiar with most of this code, though, so probably best to wait for at least one more review)

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 6 of 6 files at r5, 9 of 9 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-:-MmKUP7M7UluqVc0q8BI:bwng8lw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

**Review by knz — APPROVED** [review]
LGTM with nits

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 6 of 6 files at r5, 8 of 9 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-:-Mms4krvDARhNL8miMK7:b3nilp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/cli/clisqlshell/statement_diag.go, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-Mms3nIj1e8V54j9Xi7g:-Mms3nIj1e8V54j9Xi7h:bolr9k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a9e1b869aa045707541c5c51d18ff9383363f35/pkg/cli/clisqlshell/statement_diag.go#L75)):*
> ```Go
> 
> func (c *cliState) statementDiagList() error {
> 	const timeFmt = "2006-01-02 15:04:05 MST"
> ```

are you sure that you want to use letters to indicate the timezone? this tends to be ambiguous. why not the offset syntax?
___
*[pkg/cli/clisqlshell/statement_diag.go, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-Mms42SD5iroj9QJl3kv:-Mms42SD5iroj9QJl3kw:biocw9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a9e1b869aa045707541c5c51d18ff9383363f35/pkg/cli/clisqlshell/statement_diag.go#L83)):*
> ```Go
> 	}
> 
> 	var buf bytes.Buffer
> ```

maybe bring the decl under the `else`
___
*[pkg/cli/clisqlshell/statement_diag.go, line 94 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-Mms3uK_7Acx5DQIJHtg:-Mms3uK_7Acx5DQIJHth:b-wg6q4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a9e1b869aa045707541c5c51d18ff9383363f35/pkg/cli/clisqlshell/statement_diag.go#L94)):*
> ```Go
> 		}
> 		_ = w.Flush()
> 		fmt.Fprint(c.iCtx.stdout, buf.String())
> ```

I believe that if you pass `buf` directly, there's a special interface used to copy the contents directly to the output stream without allocting an intermediate string.

**Review by RaduBerinde — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-:-MmsSQnL60pOnnTcilML:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/clisqlshell/statement_diag.go, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-Mms3nIj1e8V54j9Xi7g:-MmsRb7C99cIBfA6vbnN:b-v0ssn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a9e1b869aa045707541c5c51d18ff9383363f35/pkg/cli/clisqlshell/statement_diag.go#L75)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

are you sure that you want to use letters to indicate the timezone? this tends to be ambiguous. why not the offset syntax?
</blockquote></details>

I copied it from other parts of the cli.  We are using `UTC().Format` so it should always show up as UTC:

```
Statement diagnostics bundles:
  ID                  Collection time          Statement
  704878944702365697  2021-10-25 17:19:42 UTC  EXPLAIN ANALYZE (DEBUG) SELECT _
  704878919639826433  2021-10-25 17:19:34 UTC  EXPLAIN ANALYZE (DEBUG) SELECT _
```
___
*[pkg/cli/clisqlshell/statement_diag.go, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-Mms42SD5iroj9QJl3kv:-MmsS46s65WNH3PVwU72:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a9e1b869aa045707541c5c51d18ff9383363f35/pkg/cli/clisqlshell/statement_diag.go#L83)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

maybe bring the decl under the `else`
</blockquote></details>

Done.
___
*[pkg/cli/clisqlshell/statement_diag.go, line 94 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71680#-Mms3uK_7Acx5DQIJHtg:-MmsS4pB2bgUpdBiRaI3:b-q6qr4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a9e1b869aa045707541c5c51d18ff9383363f35/pkg/cli/clisqlshell/statement_diag.go#L94)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I believe that if you pass `buf` directly, there's a special interface used to copy the contents directly to the output stream without allocting an intermediate string.
</blockquote></details>

Changed to `buf.WriteTo()`.

**Pull request merged**
